### PR TITLE
WIP: Improve IntervalCollection eventing API

### PR DESCRIPTION
**Pull request by Abe27342**
PR is a draft with no intent to merge as-is: it contains breaking changes. However I want to get consensus on which degrees of freedom we want to expose, and this is an attempt at augmenting it with "everything that's within reason" for discussion.

### CODE REVIEW

**Comment by Abe27342** [comment]
Notable changes from the current API surface are:

- changeInterval does not fire for property-only changes
- deleteInterval fires on both initial deletion as well as ack of that change
- changeInterval gets the previous endpoints expressed as ReferencePositions.
- changeProperties gets op/locality information just as the other ones do, but also gets the previous properties. (This one I'm a bit meh on as it requires shallow-cloning the props object, but it's not the end of the world)

I'm not sure in practice if consumers care to distinguish cases where they get the ack of their local change, and perhaps we shouldn't let them. I'd want to make sure nobody needs that before removing it though. This does differ from Amruth's suggestion of filtering the events entirely, but it's trivial to write a wrapper that filters out events with `local && op !== undefined` in the initial callback.

File: packages/dds/sequence/src/intervalCollection.ts [hunk 1]
```diff
@@ -1021,19 +1023,35 @@ export class IntervalCollectionIterator<TInterval extends ISerializableInterval>
...
      * - position change due to segment sliding (will always appear as a local change)
      * The `interval` argument reflects the new values.
+     * `previousEndpoints` contains `ReferencePosition`s at the same location as the interval's original endpoints.
+     * These references may be transient, so should only be used for location information.
+     * `local` reflects whether the change originated locally.
+     * `op` is defined if and only if the server has acked this change.
      */
     (event: "changeInterval",
-        listener: (interval: TInterval, local: boolean, op: ISequencedDocumentMessage | undefined) => void);
+        listener: (interval: TInterval, previousEndpoints: { start: ReferencePosition, end: ReferencePosition }, local: boolean, op: ISequencedDocumentMessage | undefined) => void);
```
Comment (Abe27342) [inline packages/dds/sequence/src/intervalCollection.ts:1038]: adding in the previous endpoints is a bit tricky here: for ops which explicitly change one or more points of the interval, we will just have the previous local references in hand at the time we emit the event. For sliding it's a bit harder: we'll need to plumb some previous information through some of the callbacks we have set up and then create a transient reference at the same segoff to pass to the user.

**Comment by Abe27342** [comment]
with reasonable usage of the eventing API, adding params without reordering wouldn't be a breaking change (since ts allows assignment of e.g. `() => number` to the type `(x: int) => number`). But if users are doing more crazy stuff like extending the event interface for interval collection, I think this could break things since ts only allows that with undefined and I believe it will merge multiple function declarations. Do we have any precedent for making such changes on eventing APIs? Or should I just not bother with staging any of this and throw it all into `next`?

Specifically, if we allow this type of thing then we could add the extra props that are missing in `main` and then reorder them how we like in next. all of the semantics changes on when the event is fired need to go into next.

This would mostly only matter if @amruth-ms is blocked on some of the issues filed that motivated this change. Is that still the case?

**Comment by amruth-ms** [comment]
> with reasonable usage of the eventing API, adding params without reordering wouldn't be a breaking change (since ts allows assignment of e.g. `() => number` to the type `(x: int) => number`). But if users are doing more crazy stuff like extending the event interface for interval collection, I think this could break things since ts only allows that with undefined and I believe it will merge multiple function declarations. Do we have any precedent for making such changes on eventing APIs? Or should I just not bother with staging any of this and throw it all into `next`?
> 
> Specifically, if we allow this type of thing then we could add the extra props that are missing in `main` and then reorder them how we like in next. all of the semantics changes on when the event is fired need to go into next.
> 
> This would mostly only matter if @amruth-ms is blocked on some of the issues filed that motivated this change. Is that still the case?

Hi @Abe27342 . Thank you very much for your question and the detailed comments :). The new additional parameters for `changeInterval` and `propertyChanged` event listeners look good to me. We will not see any compatibility issues if any parameters are added to the event listeners. We are using the eventing API as is and not extending the event interface. We had some designs in mind to work around the issues with eventing APIs but we have not implemented them yet. If we have changes in this PR available to us, then it would simplify quite a bit of code from our side. So yeah, it would be awesome if the changes in this PR are available to us soon.

**Comment by amruth-ms** [comment]
Hi @Abe27342. I see the comment in the notable changes comment, and it states the following

> deleteInterval fires on both initial deletion as well as ack of that change

Just wanted to check if my understanding on this statement is correct. Does this mean if a client deletes an interval, the `deleteInterval` event will get fired on the same client twice (once at initial deletion and the second after interval deletion acknowledgement)? I was assuming the client that deletes the interval will get notified upon on the initial deletion and the rest of the clients will get notified upon acknowledging the interval deletion. This change in behavior would probably not break our code but just wondering what is the advantage of firing the event twice on the same client. My apologies if I totally misunderstood the comment.

**Comment by Abe27342** [comment]
> Hi @Abe27342. I see the comment in the notable changes comment, and it states the following
> 
> > deleteInterval fires on both initial deletion as well as ack of that change
> 
> Just wanted to check if my understanding on this statement is correct. Does this mean if a client deletes an interval, the `deleteInterval` event will get fired on the same client twice (once at initial deletion and the second after interval deletion acknowledgement)? I was assuming the client that deletes the interval will get notified upon on the initial deletion and the rest of the clients will get notified upon acknowledging the interval deletion. This change in behavior would probably not break our code but just wondering what is the advantage of firing the event twice on the same client. My apologies if I totally misunderstood the comment.

You understood the comment correctly. This API isn't necessarily the final cut, though. I just proposed it as the most flexible one for the user. Whatever decision we make here, we should be consistent on whether events get fired on ack as well as initial change. So either we make each event fire on local application as well as ack (which case you're in is distinguishable from the event handler), or we forbid users from caring and take your suggestion. I'm generally ok with either, just not sure consumers will ever care about that degree of freedom.

I feel like it's overkill to expose the entire op on the event if we're not emitting the event for both cases though--it seems like any kind of logic that cared about properties on remote op application (e.g. sequence number, op contents, etc.) would require the event to be emitted as well on local ack.

**Comment by amruth-ms** [comment]
Hi @Abe27342. Thank you for your response :). Totally agree on all events being consistent. As of now are not using the "op" parameter when an interval related event is fired but that could change in the future. It would not be a problem on our side if the events get fired twice(once upon the initial change and the second after local ack) and could be useful sometime in the future. We can differentiate between them by checking if "op" is undefined or not so that we don't perform a certain action twice on our end.

**Comment by Abe27342** [comment]
#10966 is PR for implementation. I ended up tweaking semantics from as defined here (events are *not* fired on both initial change as well as ack; this is problematic for delete. Instead they're just fired once) but generally worked nicely.

**Pull request closed without merging**